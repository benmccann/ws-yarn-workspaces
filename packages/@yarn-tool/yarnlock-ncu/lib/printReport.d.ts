import { IYarnLockUpdateReport } from './types';
export declare function printReport(report: IYarnLockUpdateReport, options?: {
    chalk?: any;
    console?: any;
}): string;
