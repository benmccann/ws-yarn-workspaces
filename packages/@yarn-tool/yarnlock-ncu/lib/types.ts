
export interface IYarnLockUpdateReport
{
	removed?: Record<string, {
		from: string,
		to: string,
	}>
}

export interface IYarnLockUpdate
{
	yarnlock_old: string,
	yarnlock_new: string,
	yarnlock_changed: boolean,
	report: IYarnLockUpdateReport,
}
