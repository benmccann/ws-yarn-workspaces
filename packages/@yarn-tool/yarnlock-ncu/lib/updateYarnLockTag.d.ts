/// <reference types="node" />
import { IYarnLockUpdate } from './types';
export declare function updateYarnLockTag(yarnlock_old: Buffer | string): Promise<IYarnLockUpdate>;
