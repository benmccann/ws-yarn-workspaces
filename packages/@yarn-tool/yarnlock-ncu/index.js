"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateYarnLockTag = void 0;
const updateYarnLockTag_1 = require("./lib/updateYarnLockTag");
Object.defineProperty(exports, "updateYarnLockTag", { enumerable: true, get: function () { return updateYarnLockTag_1.updateYarnLockTag; } });
exports.default = updateYarnLockTag_1.updateYarnLockTag;
//# sourceMappingURL=index.js.map