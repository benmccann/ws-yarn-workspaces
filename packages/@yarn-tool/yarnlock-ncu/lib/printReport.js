"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.printReport = printReport;
const core_1 = require("@yarn-tool/table/lib/core");
const semver_diff_1 = require("@yarn-tool/semver-diff");
const debug_color2_1 = require("debug-color2");
const formatVersion_1 = require("@yarn-tool/yarnlock-diff/lib/formatter/formatVersion");
function printReport(report, options) {
    var _a, _b;
    // @ts-ignore
    let chalk = (_a = options === null || options === void 0 ? void 0 : options.chalk) !== null && _a !== void 0 ? _a : (0, debug_color2_1.chalkByConsoleMaybe)(options === null || options === void 0 ? void 0 : options.console);
    const table = (0, core_1.createDependencyTable)({
        colAligns: ['left', 'center', 'center', 'center'],
        head: [
            chalk.bold.reset('package'),
            chalk.bold.reset('old version'),
            '',
            chalk.bold.reset('new version'),
        ]
    });
    let formatedDiff = {};
    let _ok = false;
    const ARROW = chalk.gray('â†’');
    Object.entries((_b = report === null || report === void 0 ? void 0 : report.removed) !== null && _b !== void 0 ? _b : {})
        .forEach(([path, { from, to, }]) => {
        _ok = true;
        let lhs0 = (0, formatVersion_1._formatVersion)(from);
        let rhs0 = (0, formatVersion_1._formatVersion)(to);
        let lhs = chalk.red(lhs0);
        let rhs = chalk.yellow((0, semver_diff_1.colorizeDiff)(lhs0, rhs0, options));
        let _arr;
        _arr = [chalk.yellow(path), lhs, ARROW, rhs];
        (_arr === null || _arr === void 0 ? void 0 : _arr.length) > 0 && (formatedDiff[path] = _arr);
    });
    table.push(...Object.values(formatedDiff));
    return _ok ? table.toString() : '';
}
//# sourceMappingURL=printReport.js.map