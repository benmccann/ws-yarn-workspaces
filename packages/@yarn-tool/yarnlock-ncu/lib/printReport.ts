import { IYarnLockUpdateReport } from './types';
import { createDependencyTable } from '@yarn-tool/table/lib/core';
import { colorizeDiff, IOptionsParseVersionsDiff } from '@yarn-tool/semver-diff/index';
import { IChalk, chalkByConsoleMaybe } from 'debug-color2/index';
import { _formatVersion } from '@yarn-tool/yarnlock-diff/lib/formatter/formatVersion';

export function printReport(report: IYarnLockUpdateReport, options?: {
	chalk?,
	console?,
})
{
	// @ts-ignore
	let chalk: IChalk = options?.chalk ?? chalkByConsoleMaybe(options?.console);

	const table = createDependencyTable({
		colAligns: ['left', 'center', 'center', 'center'],
		head: [
			chalk.bold.reset('package'),
			chalk.bold.reset('old version'),
			'',
			chalk.bold.reset('new version'),
		]
	});

	let formatedDiff: {
		[k: string]: [string, string, string, string];
	} = {};

	let _ok = false;
	const ARROW = chalk.gray('â†’');

	Object.entries(report?.removed ?? {})
		.forEach(([path, {
			from,
			to,
		}]) => {

			_ok = true;

			let lhs0 = _formatVersion(from);
			let rhs0 = _formatVersion(to);

			let lhs = chalk.red(lhs0);
			let rhs = chalk.yellow(colorizeDiff(lhs0, rhs0, options));

			let _arr: [string, string, string, string];
			_arr = [chalk.yellow(path), lhs, ARROW, rhs];

			_arr?.length > 0 && (formatedDiff[path] = _arr);
		})
	;

	table.push(...Object.values(formatedDiff))

	return _ok ? table.toString() : '';
}
