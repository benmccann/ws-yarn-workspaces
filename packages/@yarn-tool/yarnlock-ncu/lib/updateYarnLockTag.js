"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateYarnLockTag = void 0;
const fromContent_1 = require("@yarn-tool/yarnlock-entries/lib/fromContent");
const queryVersion_1 = require("@yarn-tool/pkg-version-query/lib/queryVersion");
const semver_1 = require("semver");
const npm_package_arg_util_1 = require("@yarn-tool/npm-package-arg-util");
const parseSimpleSemVerRange_1 = require("@lazy-node/semver-simple-parse/lib/parseSimpleSemVerRange");
async function updateYarnLockTag(yarnlock_old) {
    const obj = (0, fromContent_1.fromContent)(yarnlock_old);
    yarnlock_old = obj.stringify();
    let report = {};
    await obj.mapAsync(async (data, key) => {
        var _a, _b;
        if (data.value.type === 'tag') {
            let { semver, version, name } = data.value;
            let version_new = await (0, queryVersion_1.queryVersionWithCache)(name, semver);
            if ((version_new === null || version_new === void 0 ? void 0 : version_new.length) && version_new !== version && (0, semver_1.gt)(version_new, version)) {
                obj.del(key);
                (_a = report.removed) !== null && _a !== void 0 ? _a : (report.removed = {});
                report.removed[key] = {
                    from: version,
                    to: version_new,
                };
            }
        }
        else if (data.value.type === 'alias') {
            let npaResult = (0, npm_package_arg_util_1.npa)(data.key);
            //			console.dir({
            //				data,
            //				npaResult,
            //			})
            if (npaResult.rawSpec.startsWith('npm:')) {
                let name = npaResult.subSpec.name;
                let version = data.value.version;
                let semver;
                let version_new;
                if (npaResult.subSpec.type === 'range') {
                    if (npaResult.subSpec.fetchSpec === "" /* EnumSemverVersion.ANY */ || npaResult.subSpec.fetchSpec === "*" /* EnumSemverVersion.STAR */) {
                        semver = '>' + version;
                    }
                    else if ((0, parseSimpleSemVerRange_1.parseSimpleSemVerRange)(npaResult.subSpec.fetchSpec).length > 1) {
                        semver = npaResult.subSpec.fetchSpec;
                    }
                    if (typeof semver === 'string') {
                        version_new = await (0, queryVersion_1.queryVersionWithCache)(name, semver);
                    }
                }
                else if (npaResult.subSpec.type === 'tag') {
                    version_new = await (0, queryVersion_1.queryVersionWithCache)(name, npaResult.subSpec.fetchSpec);
                }
                if ((version_new === null || version_new === void 0 ? void 0 : version_new.length) && version_new !== version && (0, semver_1.gt)(version_new, version)) {
                    obj.del(key);
                    (_b = report.removed) !== null && _b !== void 0 ? _b : (report.removed = {});
                    report.removed[key] = {
                        from: version,
                        to: version_new,
                    };
                }
            }
        }
    });
    let yarnlock_new = obj.stringify();
    return {
        yarnlock_old,
        yarnlock_new,
        yarnlock_changed: yarnlock_old !== yarnlock_new,
        report,
    };
}
exports.updateYarnLockTag = updateYarnLockTag;
//# sourceMappingURL=updateYarnLockTag.js.map