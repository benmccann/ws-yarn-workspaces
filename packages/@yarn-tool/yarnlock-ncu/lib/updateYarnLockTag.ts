import { fromContent } from '@yarn-tool/yarnlock-entries/lib/fromContent';
import { queryVersionWithCache } from '@yarn-tool/pkg-version-query/lib/queryVersion';
import { gt } from 'semver';
import { IYarnLockUpdateReport, IYarnLockUpdate } from './types';
import { npa } from '@yarn-tool/npm-package-arg-util';
import { AliasResult } from 'npm-package-arg';
import { EnumSemverVersion } from '@lazy-node/semver-ampersand/lib/const';
import { parseSimpleSemVerRange } from '@lazy-node/semver-simple-parse/lib/parseSimpleSemVerRange';

export async function updateYarnLockTag(yarnlock_old: Buffer | string)
{
	const obj = fromContent(yarnlock_old);

	yarnlock_old = obj.stringify();

	let report: IYarnLockUpdateReport = {};

	await obj.mapAsync(async (data, key) => {

		if (data.value.type === 'tag')
		{
			let { semver, version, name } = data.value;

			let version_new = await queryVersionWithCache(name, semver);

			if (version_new?.length && version_new !== version && gt(version_new, version))
			{
				obj.del(key);

				report.removed ??= {};

				report.removed[key] = {
					from: version,
					to: version_new,
				};
			}

		}
		else if (data.value.type === 'alias')
		{
			let npaResult = npa<AliasResult>(data.key);

//			console.dir({
//				data,
//				npaResult,
//			})

			if (npaResult.rawSpec.startsWith('npm:'))
			{
				let name = npaResult.subSpec.name;
				let version = data.value.version;
				let semver: string;
				let version_new: string;

				if (npaResult.subSpec.type === 'range')
				{
					if (npaResult.subSpec.fetchSpec === EnumSemverVersion.ANY || npaResult.subSpec.fetchSpec === EnumSemverVersion.STAR)
					{
						semver = '>' + version;
					}
					else if (parseSimpleSemVerRange(npaResult.subSpec.fetchSpec).length > 1)
					{
						semver = npaResult.subSpec.fetchSpec;
					}

					if (typeof semver === 'string')
					{
						version_new = await queryVersionWithCache(name, semver);
					}
				}
				else if (npaResult.subSpec.type === 'tag')
				{
					version_new = await queryVersionWithCache(name, npaResult.subSpec.fetchSpec);
				}

				if (version_new?.length && version_new !== version && gt(version_new, version))
				{
					obj.del(key);

					report.removed ??= {};

					report.removed[key] = {
						from: version,
						to: version_new,
					};
				}
			}
		}

	});

	let yarnlock_new = obj.stringify();

	return {
		yarnlock_old,
		yarnlock_new,
		yarnlock_changed: yarnlock_old !== yarnlock_new,
		report,
	} as IYarnLockUpdate
}
