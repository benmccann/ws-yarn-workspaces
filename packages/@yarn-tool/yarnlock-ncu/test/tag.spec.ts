import { join } from "path";
import { updateYarnLockTag } from '../lib/updateYarnLockTag';
import { readFileSync } from "fs";
import { __TEST_YARNLOCK } from '../../../../__root_ws';
import { printReport } from '../lib/printReport';
import stripAnsi from 'strip-ansi';
import { pathExistsSync } from 'fs-extra';
import { EnumDetectYarnLock } from '@yarn-tool/yarnlock-types';
import { _forEachVersionTags } from '../../../../test/lib/forEachVersionTags';

const TIMEOUT = 5 * 60 * 1000;

describe(`fixYarnLockTagUpdate`, () =>
{
	_forEachVersionTags().forEach(ver =>
	{
		describe(ver, () =>
		{
			const dir = join(__TEST_YARNLOCK, 'ncu', ver);

			if (!pathExistsSync(dir))
			{
				test(`dummy`, () => {})
			}
			else
			{
				test(`should update tag in yarn.lock`, async () =>
				{
					const file = join(dir, 'yarn.lock');
					let buf = readFileSync(file);

					let actual = await updateYarnLockTag(buf);

					console.dir(actual.yarnlock_changed);

					expect(actual).toHaveProperty(['yarnlock_changed'], true);

					console.dir(actual.report);

					expect(actual).toHaveProperty(['report', 'removed', 'next@canary']);
					expect(actual).toHaveProperty(['report', 'removed', 'typescript@next']);

					expect(actual).toMatchSnapshot({
						yarnlock_new: expect.any(String),
						yarnlock_old: expect.any(String),
						report: {
							removed: {
								'next@canary': {
									to: expect.any(String),
								},
								'typescript@next': {
									to: expect.any(String),
								},
							},
						},
					});

					console.log(printReport(actual.report));

				}, TIMEOUT);
			}

		});
	});

})
