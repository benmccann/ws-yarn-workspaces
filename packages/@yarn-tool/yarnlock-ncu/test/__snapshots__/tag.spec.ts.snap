// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixYarnLockTagUpdate v1 should update tag in yarn.lock 1`] = `
Object {
  "report": Object {
    "removed": Object {
      "next@canary": Object {
        "from": "9.5.2-canary.1",
        "to": Any<String>,
      },
      "typescript@next": Object {
        "from": "4.0.0-dev.20200731",
        "to": Any<String>,
      },
    },
  },
  "yarnlock_changed": true,
  "yarnlock_new": Any<String>,
  "yarnlock_old": Any<String>,
}
`;

exports[`fixYarnLockTagUpdate v3 should update tag in yarn.lock 1`] = `
Object {
  "report": Object {
    "removed": Object {
      "next@canary": Object {
        "from": "9.5.2-canary.1",
        "to": Any<String>,
      },
      "typescript@next": Object {
        "from": "4.0.0-dev.20200731",
        "to": Any<String>,
      },
    },
  },
  "yarnlock_changed": true,
  "yarnlock_new": Any<String>,
  "yarnlock_old": Any<String>,
}
`;
